File: app/src/main/java/br/com/dbccompany/sicred/presentation/event/ui/dialog/CheckInViewModel.java
package br.com.dbccompany.sicred.presentation.event.ui.dialog;

import androidx.hilt.Assisted;
import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;

import br.com.dbccompany.sicred.data.datasource.remote.checkin.CheckInResponse;
import br.com.dbccompany.sicred.data.repositories.EventRepository;
import br.com.dbccompany.sicred.domain.model.CheckIn;
import br.com.dbccompany.sicred.domain.model.Event;
import br.com.dbccompany.sicred.domain.usecase.RealizeCheckinUseCase;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CheckInViewModel extends ViewModel {

    private final RealizeCheckinUseCase realizeCheckInUseCase;
    private final SavedStateHandle savedStateHandle;
    private final EventRepository eventRepository;
    private MutableLiveData<Boolean> showProgress = new MutableLiveData<>();
    private MutableLiveData<Boolean> showError = new MutableLiveData<>();
    private MutableLiveData<Boolean> showContent = new MutableLiveData<>();
    private String eventId;
    private MutableLiveData<Boolean> showSuccess = new MutableLiveData<>();

    @ViewModelInject
    public CheckInViewModel(EventRepository eventRepository,
                            @Assisted SavedStateHandle savedStateHandle,
                            RealizeCheckinUseCase realizeCheckinUseCase) {
        this.eventRepository = eventRepository;
        this.savedStateHandle = savedStateHandle;
        this.realizeCheckInUseCase = realizeCheckinUseCase;
    }

    public void setErrorState(){
        showError.postValue(true);
        showProgress.postValue(false);
        showContent.postValue(false);
    }
    private void setShowSuccess() {
        showContent.postValue(true);
        showSuccess.postValue(true);
        showProgress.postValue(false);
        showError.postValue(false);
    }
    public void setLoadState(){
        showProgress.postValue(true);
        showError.postValue(false);
        showContent.postValue(false);
    }

    public void checkInAction(String name, String email, String eventId) {
        setLoadState();
        realizeCheckInUseCase.invoke(new CheckIn(eventId, name, email))
                .enqueue(new Callback<CheckInResponse>() {
            @Override
            public void onResponse(Call<CheckInResponse> call, Response<CheckInResponse> response) {
                    setShowSuccess();
            }

            @Override
            public void onFailure(Call<CheckInResponse> call, Throwable t) {
                setErrorState();
            }
        });
    }

    public void setEventId(String eventId) {
        this.eventId = eventId;
    }

    public String getEventId() {
        return eventId;
    }

    public RealizeCheckinUseCase getRealizeCheckInUseCase() {
        return realizeCheckInUseCase;
    }

    public SavedStateHandle getSavedStateHandle() {
        return savedStateHandle;
    }

    public EventRepository getEventRepository() {
        return eventRepository;
    }

    public MutableLiveData<Boolean> getShowProgress() {
        return showProgress;
    }

    public void setShowProgress(MutableLiveData<Boolean> showProgress) {
        this.showProgress = showProgress;
    }

    public MutableLiveData<Boolean> getShowError() {
        return showError;
    }

    public void setShowError(MutableLiveData<Boolean> showError) {
        this.showError = showError;
    }

    public MutableLiveData<Boolean> getShowContent() {
        return showContent;
    }

    public void setShowContent(MutableLiveData<Boolean> showContent) {
        this.showContent = showContent;
    }

    public MutableLiveData<Boolean> getShowSuccess() {
        return showSuccess;
    }

    public void setShowSuccess(MutableLiveData<Boolean> showSuccess) {
        this.showSuccess = showSuccess;
    }
}

File: app/src/main/java/br/com/diogomoreno/events/presentation/event/ui/dialog/CheckInViewModel.kt
package br.com.diogomoreno.events.presentation.event.ui.dialog

import androidx.hilt.Assisted
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import br.com.diogomoreno.events.data.datasource.remote.checkin.CheckInResponse
import br.com.diogomoreno.events.data.repositories.EventRepository
import br.com.diogomoreno.events.domain.model.CheckIn
import br.com.diogomoreno.events.domain.usecase.RealizeCheckinUseCase
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class CheckInViewModel @ViewModelInject constructor(val eventRepository: EventRepository,
                                                    @param:Assisted val savedStateHandle: SavedStateHandle,
                                                    val realizeCheckInUseCase: RealizeCheckinUseCase) : ViewModel() {
    var showProgress = MutableLiveData<Boolean>()
    var showError = MutableLiveData<Boolean>()
    var showContent = MutableLiveData<Boolean>()
    var eventId: String? = null
    var showSuccess = MutableLiveData<Boolean>()
    fun setErrorState() {
        showError.postValue(true)
        showProgress.postValue(false)
        showContent.postValue(false)
    }

    private fun setShowSuccess() {
        showContent.postValue(true)
        showSuccess.postValue(true)
        showProgress.postValue(false)
        showError.postValue(false)
    }

    fun setLoadState() {
        showProgress.postValue(true)
        showError.postValue(false)
        showContent.postValue(false)
    }

    fun checkInAction(name: String?, email: String?, eventId: String?) {
        setLoadState()
        realizeCheckInUseCase.invoke(CheckIn(eventId, name, email))
                ?.enqueue(object : Callback<CheckInResponse?> {
                    override fun onResponse(call: Call<CheckInResponse?>, response: Response<CheckInResponse?>) {
                        setShowSuccess()
                    }

                    override fun onFailure(call: Call<CheckInResponse?>, t: Throwable) {
                        setErrorState()
                    }
                })
    }

}